File: c_solutions/9-palindrome-number/palindrome-number.py
#git sync

class Solution:
    def isPalindrome(self, x: int) -> bool:
        # Negative numbers are not palindromes
        if x < 0:
            return False
        
        # Store the original number to compare later
        original = x
        reversed_num = 0
        
        while x > 0:
            digit = x % 10  # Get the last digit
            reversed_num = reversed_num * 10 + digit  # Build reversed number
            x //= 10  # Remove the last digit
        
        return original == reversed_num
        

File: c_solutions/9-palindrome-number/palindrome-number.c
bool isPalindrome(int x) {
    // Negative numbers are not palindromes
    if (x < 0) return false;
    
    // Store the original number
    int original = x, reversed = 0;
    
    while (x > 0) {
        int digit = x % 10;  // Extract last digit
        if (reversed > (2147483647 - digit) / 10) return false; // Prevent overflow
        reversed = reversed * 10 + digit; // Build reversed number
        x /= 10;  // Remove last digit
    }
    
    return original == reversed;
}